File: Graphs/Course Schedule/ans2.py
class Solution(object):
    def dfs(self, i, edges, visited, st):
        st[i] = 1
        if visited[i] == 0:
            visited[i] = 1
            for x in edges[i]:
                if visited[x] == 0 and self.dfs(x, edges, visited, st):
                    return True
                if st[x] == 1:
                    return True
        
        st[i] = 0
        return False

    def canFinish(self, n, pre):
        edges = defaultdict(list)
        for x in pre:
            edges[x[1]].append(x[0])
        
        visited = [0]*n
        st = [0]*n

        cycle = False
        for i in range(n):
            if visited[i] == 0 and self.dfs(i, edges, visited, st):
                cycle = True
                break
        
        return not cycle

File: Graphs/Course Schedule/ans2.cpp
class Solution {
public:
    bool dfs(vector<vector<int>> &edges, vector<int> &visited, vector<int> &st, int i){
        st[i] = 1;
        if(!visited[i]){
           visited[i] = true;
           for(auto x : edges[i]){
               if(!visited[x] && dfs(edges, visited, st, x)){
                   return true;
               }
               if(st[x] == 1) return true;
           }
        }
        st[i] = false;
        return false;
    }
    bool canFinish(int courses, vector<vector<int>>& pre) {
        vector<vector<int>> edges(courses);
        for(auto it : pre){
            edges[it[1]].push_back(it[0]);
        }
        vector<int> visited(courses, 0), st(courses, 0);
        bool cycle = false;

        for(int i = 0; i<courses; i++){
           if(!visited[i] && dfs(edges, visited, st, i)) cycle = true;
        }

        return !cycle;
    }
};